### ISSUE TITLE
Need a way to preserve initial indention of the input expression

### ISSUE THREAD

**Issue by renkun-ken** [r-lib/styler#641]
In a code editor, when user performs range formatting, the selected code might have initial indention. The code editor should preserve the indention since the selected code could be part of nested scopes like the following:

No initial indention:

```r
data <- dbQuery(
  con,
  "select name, group, max(update_time) as max_time
     from users
     where date = 20200525
     group by group"
)
```

`data <- dbQuery(...)` has initial indention to be preserved:

```r
fun <- function(x, y) {
  data <- dbQuery(
    con,
    "select name, group, max(update_time) as max_time
     from users
     where date = 20200525
     group by group"
  )
}
```

![image](https://user-images.githubusercontent.com/4662568/82775638-e0076080-9e7a-11ea-923a-9f636b041835.png)

Calling `styler::style_text()` on text with initial indention removes it.

I'm wondering if there's a simple way to preserve the initial indention of the input code?

Related to https://github.com/REditorSupport/languageserver/issues/264.

**Comment by lorenzwalthert** [comment]
We have the same problem with the {styler} Addin to style the active selection. One workaround is just to always style whole expressions. I never use this feature because I use {precommit} which styles the whole file, i.e. I don't really see the point of styling only one part of the file. Also note that because the way styler works, we assume that the first line should never have leading spaces, even for `scope < 'indention'`. Hence, the way to go about it is IMO to 

1. infer the number of leading spaces from the first line of code passed to styler.
2. `paste0(paste0(rep(" ", n_leading_spaces), collapse = ""), style_text(input_text)`.

This requires that the highlighted selection starts at column 1, i.e. as in your screenshot above, and not like this:
![Screenshot 2020-05-25 at 08 33 32](https://user-images.githubusercontent.com/10477073/82785214-6c595980-9e62-11ea-860b-8a0ce189ccc0.png)

For the latter case, you'd need to check that all content on the first selected line before the highlighted area does not contain any token to safely take this as the current indention.

**Comment by renkun-ken** [comment]
@lorenzwalthert Then there's https://github.com/REditorSupport/languageserver/issues/264 where the initial indention is also added to multi-line string literals, which is undesired.

**Comment by lorenzwalthert** [comment]
Uh, sorry, missed that.

**Comment by renkun-ken** [comment]
In https://github.com/REditorSupport/languageserver/pull/268, changing top-level expression `indent` does not work, so I use a trick of enclosing the expressions into `{ ... }` to resolve this, assuming that `scope >= "indention"`.

It would be nice if the top-level expression could support `indent` somehow.

**Comment by lorenzwalthert** [comment]
Can you check if `scope >= 'indention'` in your code or is that an implicit assumption you make? Can you detail how such an argument would work? I.e. do you say "indent = 2" and then all code is indented by 2 (ensuring multi-line string linterals don't get re-indented)?

**Comment by renkun-ken** [comment]
Yes, what is wanted is just adding `n` spaces to all expressions except for multi-line strings. I add a check to ensure it only applies when `scope >= "indention"`.

It would be nice if there's built-in support for global indention like this.

**Comment by lorenzwalthert** [comment]
Fixed @renkun-ken, hope you can adapt now accordingly in REditorSupport/languageserver.

**Comment by renkun-ken** [comment]
Thanks @lorenzwalthert! I'll try to adapt this in languageserver.

**Comment by renkun-ken** [comment]
@lorenzwalthert I tried the `base_indention` argument, and it seems that it has some inconsistent behavior. For example,

```
> styler::style_text(c("x<-1", "y<2"), base_indention = 4)                                                       
    x <- 1
    y < 2

> styler::style_text(c("x<-1", "0"), base_indention = 4)                                                         
    x <- 1
0
```

**Comment by lorenzwalthert** [comment]
Right... The problem is caused by the cache only though,i.e. when you turn it off, it won't fail. Need to look into it again, thanks for reporting.

**Comment by lorenzwalthert** [comment]
Tracked in #655. How urgent is this @renkun-ken?

**Comment by renkun-ken** [comment]
Take your time. Not urgent at all.

**Comment by lorenzwalthert** [comment]
Closed with #655.

**Comment by lorenzwalthert** [comment]
@renkun-ken please try it out and feel free to re-open if problem persists.
